### ISSUE TITLE
Java, Windows, MinGW GCC compiler issue

### ISSUE THREAD

**Issue by Nashatyrev** [swig/swig#1932]
Environment: Windows, MinGW 64-bit GCC compiler, cygwin or msys2 shell.
The original issue: https://github.com/supranational/blst/issues/53

When generating wrapper cpp code we've got the following fragment in our generated wrapper cpp file 
```
/* Fix for jlong on some versions of gcc on Windows */
#if defined(__GNUC__) && !defined(__INTEL_COMPILER)
  typedef long long __int64;
#endif
```
The above code comes from the SWIG [`javahead.swg` include](https://github.com/swig/swig/blob/88896356831a025220b6ebffcb38f0ac08dcddef/Lib/java/javahead.swg#L32-L36)

When trying to compile the generated code with MinGW-64 GCC compiler it fails with: 
```
blst_wrap.cpp:194:21: error: 'long long long' is too long for GCC
  194 |   typedef long long __int64;
```

This is because MinGW GCC has the header included `_mingw.h` which does `#define __int64 long long` which obviously breaks the `typedef` above. Below is the inclusion path: 
```
. ../blst.hpp
.. E:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/9.2.0/include/c++/string
... E:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/9.2.0/include/c++/bits/char_traits.h
.... E:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/9.2.0/include/c++/bits/postypes.h
..... E:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/9.2.0/include/c++/cwchar
...... E:/TDM-GCC-64/x86_64-w64-mingw32/include/wchar.h
....... E:/TDM-GCC-64/x86_64-w64-mingw32/include/corecrt.h
........ E:/TDM-GCC-64/x86_64-w64-mingw32/include/_mingw.h
```

May be it makes sense to add `&& !defined(__int64)` to the macro condition? (I'm not a big C specialist though)

**Comment by Nashatyrev** [comment]
One clarification which might matter in this context: we are using TDM-GCC distribution. See [this comment](https://github.com/supranational/blst/issues/53#issuecomment-764966887) for a bit more details

**Comment by Nashatyrev** [comment]
Here is the compiler details: 
```
$ gcc -v
Using built-in specs.
COLLECT_GCC=E:\TDM-GCC-64\bin\gcc.exe
COLLECT_LTO_WRAPPER=E:/TDM-GCC-64/bin/../libexec/gcc/x86_64-w64-mingw32/9.2.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../../../src/gcc-git-9.2.0/configure --build=x86_64-w64-mingw32 --enable-targets=all --enable-languages=ada,c,c++,fortran,lto,objc,obj-c
++ --enable-libgomp --enable-lto --enable-graphite --enable-cxx-flags=-DWINPTHREAD_STATIC --disable-build-with-cxx --disable-build-poststage1-with-cxx --
enable-libstdcxx-debug --enable-threads=posix --enable-version-specific-runtime-libs --enable-fully-dynamic-string --enable-libstdcxx-threads --enable-li
bstdcxx-time --with-gnu-ld --disable-werror --disable-nls --disable-win32-registry --enable-large-address-aware --disable-rpath --disable-symvers --prefi
x=/mingw64tdm --with-local-prefix=/mingw64tdm --with-pkgversion=tdm64-1 --with-bugurl=http://tdm-gcc.tdragon.net/bugs
Thread model: posix
gcc version 9.2.0 (tdm64-1)
```

**Comment by Nashatyrev** [comment]
Here is also the workaround on the client side which does the trick: https://github.com/supranational/blst/issues/53#issuecomment-764709490

**Comment by dot-asm** [comment]
The referred `typedef` in generated wrapper is formally problematic, because it effectively makes an assumption that no headers are included in `%begin` section. Well, not exactly "no headers," but you're likely to have hard time finding one that wouldn't trigger the problem. Either way, such assumption is formally inappropriate to make. While it was possible to mitigate the problem in [blst,](https://github.com/supranational/blst/commit/a4694fd8ab64c3182b9428eb7228c85f81299fe1) the workaround is dependent on SWIG doing this formally problematic thing. More specifically. More appropriate thing for SWIG to do is to replace
```
/* Fix for jlong on some versions of gcc on Windows */
#if defined(__GNUC__) && !defined(__INTEL_COMPILER)
  typedef long long __int64;
#endif
```
with
```
#include <stdint.h>
```
But in combination with the current workaround such modification would trigger compilation error of a different kind. So, workaround works, but it might be rendered bad if SWIG chooses to apply more appropriate solution.

Just in case for reference. Even though stdint.h is meant to instantiate things like `int<n>_t` and `uint<n>_t`, MinGW header is bound to define even Visual-C-specific `__int64`. If unconditional inclusion of stdint.h feels less appropriate it's possible to
```
#ifdef __MINGW32__
#include <stdint.h>
#endif
```
`__MINGW32__` is defined in **both** 32- and 64-bit compilations (while `__MINGW64__` is defined only in 64-bit compilation).

**Comment by dot-asm** [comment]
Actually having a bit broader look, one can as well advocate for replacing whole
```
/* Fix for jlong on some versions of gcc on Windows */
#if defined(__GNUC__) && !defined(__INTEL_COMPILER)
  typedef long long __int64;
#endif

/* Fix for jlong on 64-bit x86 Solaris */
#if defined(__x86_64)
# ifdef _LP64
#   undef _LP64
# endif
#endif

#include <jni.h>
#include <stdlib.h>
#include <string.h>
```
with just
```
#include <stdlib.h>
#include <string.h>
#include <jni.h>
```
Just in case, as for omission of `#undef _LP64`. Undefining **pre-defined** macro, which `_LP64` is, is **extremely** bad style. At the very least it's worse than assuming that no headers are included in `%begin` section:-) Commentary claims it was some problem on Solaris. I'd be inclined to believe that it was a misunderstanding of some kind, but even if the problem was real, why did it qualify for universal `#undef`? After all, `_LP64` is not Solaris-specific macro...

**Comment by dot-asm** [comment]
@Nashatyrev, this is resolved, please, close:-)

**Comment by wsfulton** [comment]
Fixed by #1934, thanks @dot-asm.
